File: scripts/enrich-municipality-contacts.py
#!/usr/bin/env python3
"""
Enrich US municipalities (100k-500k pop) with government contact data via Apollo.io.

Two-step flow:
  1. POST /api/v1/mixed_people/api_search — find candidate people at the org
  2. POST /api/v1/people/match — unlock real name + verified email (costs credits)

Checkpoint file at scripts/.enrich-progress.json lets the run resume after interruption.
"""

import json
import os
import sys
import time
from pathlib import Path

import requests

ROOT = Path(__file__).resolve().parent.parent
DATA_FILE = ROOT / "public" / "arsenal-municipality-data.json"
PROGRESS_FILE = Path(__file__).resolve().parent / ".enrich-progress.json"

APOLLO_KEY = os.environ.get("APOLLO_API_KEY", "").strip()
if not APOLLO_KEY:
    print("ERROR: APOLLO_API_KEY env var required", file=sys.stderr)
    sys.exit(1)

SEARCH_URL = "https://api.apollo.io/api/v1/mixed_people/api_search"
MATCH_URL = "https://api.apollo.io/api/v1/people/match"
HEADERS = {
    "Content-Type": "application/json",
    "accept": "application/json",
    "Cache-Control": "no-cache",
    "X-Api-Key": APOLLO_KEY,
}

TITLE_PRIORITY = [
    "city manager", "town manager", "village manager", "borough manager",
    "city administrator", "town administrator", "village administrator", "borough administrator",
    "mayor",
    "city clerk", "town clerk", "village clerk", "borough clerk",
    "community development director", "economic development director",
    "public works director", "finance director",
    "city attorney", "town attorney",
]

TYPE_DISPLAY = {
    "city": "City",
    "town": "Town",
    "village": "Village",
    "borough": "Borough",
    "city and borough": "City and Borough",
    "unified government": "Unified Government",
    "consolidated government": "Consolidated Government",
    "urban county": "Urban County",
    "metropolitan government": "Metropolitan Government",
    "plantation": "Plantation",
    "municipality": "Municipality",
}


def format_org(place):
    name = place["name"]
    for suffix in (" city", " town", " village", " borough", " municipality"):
        if name.lower().endswith(suffix):
            name = name[: -len(suffix)].strip()
            break
    type_display = TYPE_DISPLAY.get(place.get("type", ""), "City")
    return f"{type_display} of {name}"


def load_progress():
    if PROGRESS_FILE.exists():
        try:
            return json.loads(PROGRESS_FILE.read_text())
        except Exception:
            pass
    return {"processed": [], "enriched": 0, "attempted": 0}


def save_progress(progress):
    PROGRESS_FILE.write_text(json.dumps(progress, indent=2))


def search(org_name, state_abbr):
    body = {
        "q_organization_name": org_name,
        "person_titles": TITLE_PRIORITY,
        "organization_locations": [state_abbr],
        "page": 1,
        "per_page": 10,
    }
    try:
        r = requests.post(SEARCH_URL, headers=HEADERS, json=body, timeout=30)
    except Exception as e:
        print(f"    search error: {e}")
        return []
    if r.status_code != 200:
        print(f"    search {r.status_code}: {r.text[:200]}")
        return []
    return r.json().get("people", [])


def match(person_id):
    try:
        r = requests.post(
            MATCH_URL,
            headers=HEADERS,
            json={"id": person_id, "reveal_personal_emails": True},
            timeout=30,
        )
    except Exception as e:
        print(f"    match error: {e}")
        return None
    if r.status_code != 200:
        print(f"    match {r.status_code}: {r.text[:200]}")
        return None
    return r.json().get("person")


def best_candidate(people, org_name_lower):
    # Prefer candidates whose organization matches our target
    # and whose title matches our priority list
    priority_titles = [t.lower() for t in TITLE_PRIORITY]
    scored = []
    for p in people:
        org = (p.get("organization") or {}).get("name", "").lower()
        title = (p.get("title") or "").lower()
        score = 0
        if org and org_name_lower in org:
            score += 10
        for i, pt in enumerate(priority_titles):
            if pt in title:
                score += 100 - i  # earlier = higher
                break
        scored.append((score, p))
    scored.sort(key=lambda x: x[0], reverse=True)
    return [p for score, p in scored if score > 0]


def main():
    print("Loading municipality data...")
    data = json.loads(DATA_FILE.read_text())
    places = data["places"]
    print(f"Total municipalities: {len(places):,}")

    targets = [p for p in places if (p.get("population_2020") or 0) and 100000 <= p["population_2020"] <= 500000]
    print(f"Target (100k-500k): {len(targets):,}")

    progress = load_progress()
    todo = [p for p in targets if p["geoid"] not in set(progress["processed"])]
    print(f"To process: {len(todo):,} ({len(progress['processed'])} already processed)")

    if not todo:
        print("All targets already processed.")
        return

    start = time.time()
    by_geoid = {p["geoid"]: p for p in places}

    for i, place in enumerate(todo, 1):
        geoid = place["geoid"]
        name = place["name"]
        state = place["state_abbr"]
        pop = place.get("population_2020", 0)
        org_name = format_org(place)
        print(f"[{i}/{len(todo)}] {name}, {state} ({pop:,}) — searching '{org_name}'")

        candidates = search(org_name, state)
        ranked = best_candidate(candidates, org_name.lower())

        picked = None
        if ranked:
            top = ranked[0]
            person = match(top["id"])
            if person and person.get("email"):
                picked = person

        target = by_geoid[geoid]
        if picked:
            full_name = f"{picked.get('first_name','')} {picked.get('last_name','')}".strip()
            target["contact_name"] = full_name
            target["contact_title"] = picked.get("title") or ""
            target["contact_email"] = picked.get("email") or ""
            target["contact_phone"] = (
                picked.get("phone_numbers", [{}])[0].get("sanitized_number", "")
                if picked.get("phone_numbers")
                else ""
            )
            target["contact_source"] = "apollo"
            target["contact_enriched_at"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
            progress["enriched"] += 1
            print(f"    ✓ {full_name} — {target['contact_email']}")
        else:
            target["contact_enriched_at"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
            target["contact_source"] = "apollo_attempted"
            print(f"    ✗ no verified email found")

        progress["processed"].append(geoid)
        progress["attempted"] += 1
        save_progress(progress)

        # Persist JSON every 10 items so a crash doesn't lose everything
        if i % 10 == 0:
            DATA_FILE.write_text(json.dumps(data, indent=2))

        # Gentle pacing
        time.sleep(0.5)

    # Final write
    if "metadata" not in data:
        data["metadata"] = {}
    data["metadata"]["enrichment"] = {
        "last_run": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "target_population_band": "100k-500k",
        "target_count": len(targets),
        "processed_count": progress["attempted"],
        "enriched_count": progress["enriched"],
        "source": "Apollo.io api_search + people/match",
    }
    DATA_FILE.write_text(json.dumps(data, indent=2))

    elapsed = time.time() - start
    print()
    print(f"Complete in {elapsed/60:.1f} min")
    print(f"  Attempted: {progress['attempted']}")
    print(f"  Enriched:  {progress['enriched']}")
    print(f"  Hit rate:  {progress['enriched']/progress['attempted']*100:.1f}%")


if __name__ == "__main__":
    main()

File: scripts/enrich-municipality-contacts.js
#!/usr/bin/env node

/**
 * enrich-municipality-contacts.js
 *
 * Enriches municipalities (50k+ population) with contact data from Apollo.io
 * 
 * Steps:
 * 1. Filter 19,475 records to municipalities with 50k+ population
 * 2. Format each municipality name into an org search term
 * 3. Hit Apollo's People Search API with title filters
 * 4. Write matched contacts back onto the original JSON records
 * 5. Use checkpoint file for re-runs
 * 
 * Run: node scripts/enrich-municipality-contacts.js
 * Requires: APOLLO_API_KEY environment variable
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, '..');

// Configuration
const DATA_FILE = path.join(ROOT, 'public', 'arsenal-municipality-data.json');
const PROGRESS_FILE = path.join(__dirname, '.enrich-progress.json');
const APOLLO_API_KEY = process.env.APOLLO_API_KEY;

if (!APOLLO_API_KEY) {
  console.error('❌ APOLLO_API_KEY environment variable is required');
  console.error('   Set it: export APOLLO_API_KEY=your_key_here');
  process.exit(1);
}

// Title filters for municipal government roles
const TITLE_FILTERS = [
  'city manager',
  'town manager',
  'village manager',
  'borough manager',
  'mayor',
  'city administrator',
  'town administrator',
  'village administrator',
  'borough administrator',
  'city clerk',
  'town clerk',
  'village clerk',
  'borough clerk',
  'city planner',
  'town planner',
  'village planner',
  'borough planner',
  'economic development director',
  'community development director',
  'public works director',
  'finance director',
  'city attorney',
  'town attorney',
  'village attorney',
  'borough attorney'
];

// Format municipality name for Apollo search
function formatOrgSearchTerm(municipality) {
  const { name, type } = municipality;
  
  // Remove "city", "town", "village", "borough" from name if present
  // Handle both "Chicago city" and "city of Chicago" patterns
  let cleanName = name
    .replace(/\s+(city|town|village|borough|municipality)$/i, '')
    .replace(/^(city|town|village|borough|municipality)\s+of\s+/i, '')
    .replace(/^the\s+/i, '')
    .trim();
  
  // Format as "City of {Name}" or appropriate type
  const typeMap = {
    'city': 'City',
    'town': 'Town', 
    'village': 'Village',
    'borough': 'Borough',
    'city and borough': 'City and Borough',
    'unified government': 'Unified Government',
    'consolidated government': 'Consolidated Government',
    'urban county': 'Urban County',
    'metropolitan government': 'Metropolitan Government',
    'plantation': 'Plantation',
    'municipality': 'Municipality'
  };
  
  const typeDisplay = typeMap[type] || 'Municipality';
  return `${typeDisplay} of ${cleanName}`;
}

// Apollo People Search API call (two-step: search then bulk_match to reveal)
async function searchApolloContacts(orgName, stateAbbr) {
  const APOLLO_HEADERS = {
    'Content-Type': 'application/json',
    'Cache-Control': 'no-cache',
    'X-Api-Key': APOLLO_API_KEY
  };

  try {
    // Step 1: search for people (returns obfuscated data + IDs)
    const searchResp = await fetch('https://api.apollo.io/api/v1/mixed_people/api_search', {
      method: 'POST',
      headers: APOLLO_HEADERS,
      body: JSON.stringify({
        q_organization_name: orgName,
        organization_locations: [stateAbbr],
        page: 1,
        per_page: 10,
        person_titles: TITLE_FILTERS
      })
    });

    if (!searchResp.ok) {
      console.error(`  Apollo search error: ${searchResp.status} ${searchResp.statusText}`);
      return [];
    }

    const searchData = await searchResp.json();
    if (!searchData.people || !Array.isArray(searchData.people) || searchData.people.length === 0) {
      return [];
    }

    // Step 2: bulk_match to reveal full names and emails
    const ids = searchData.people.map(p => ({ id: p.id }));
    const matchResp = await fetch('https://api.apollo.io/api/v1/people/bulk_match', {
      method: 'POST',
      headers: APOLLO_HEADERS,
      body: JSON.stringify({ reveal_personal_emails: true, details: ids })
    });

    if (!matchResp.ok) {
      console.error(`  Apollo bulk_match error: ${matchResp.status} ${matchResp.statusText}`);
      return [];
    }

    const matchData = await matchResp.json();
    if (!matchData.matches || !Array.isArray(matchData.matches)) {
      return [];
    }

    return matchData.matches
      .filter(person => {
        const email = person.email;
        return email && !email.includes('***') && email.includes('@');
      })
      .map(person => ({
        name: person.name || `${person.first_name || ''} ${person.last_name || ''}`.trim(),
        title: person.title || '',
        email: person.email || '',
        phone: (person.phone_numbers && person.phone_numbers[0]?.sanitized_number) || person.phone || '',
        linkedin_url: person.linkedin_url || '',
        source: 'apollo'
      }));

  } catch (error) {
    console.error(`  Apollo search failed: ${error.message}`);
    return [];
  }
}

// Load progress checkpoint
function loadProgress() {
  try {
    if (fs.existsSync(PROGRESS_FILE)) {
      const data = fs.readFileSync(PROGRESS_FILE, 'utf8');
      return JSON.parse(data);
    }
  } catch (error) {
    console.warn(`Could not load progress file: ${error.message}`);
  }
  
  return {
    processed: [],
    lastRun: null,
    totalFound: 0,
    totalEnriched: 0
  };
}

// Save progress checkpoint
function saveProgress(progress) {
  try {
    fs.writeFileSync(PROGRESS_FILE, JSON.stringify(progress, null, 2));
  } catch (error) {
    console.error(`Failed to save progress: ${error.message}`);
  }
}

// Main execution
async function main() {
  console.log('📊 Loading municipality data...');
  
  // Load data
  const rawData = fs.readFileSync(DATA_FILE, 'utf8');
  const data = JSON.parse(rawData);
  const places = data.places;
  
  console.log(`📊 Total municipalities: ${places.length.toLocaleString()}`);
  
  // Filter to 50k+ population
  const targetPlaces = places.filter(place => {
    const population = place.population_2020;
    return population >= 50000;
  });
  
  console.log(`🎯 Target municipalities (50k+): ${targetPlaces.length.toLocaleString()}`);
  
  // Load progress
  const progress = loadProgress();
  progress.lastRun = new Date().toISOString();
  
  // Filter out already processed
  const toProcess = targetPlaces.filter(place => 
    !progress.processed.includes(place.geoid)
  );
  
  console.log(`⚡ To process: ${toProcess.length.toLocaleString()} (${progress.processed.length} already processed)`);
  
  if (toProcess.length === 0) {
    console.log('✅ All target municipalities already processed.');
    return;
  }
  
  // Process in batches with delay to respect rate limits
  const BATCH_SIZE = 5;
  const DELAY_MS = 2000; // 2 seconds between batches
  
  for (let i = 0; i < toProcess.length; i += BATCH_SIZE) {
    const batch = toProcess.slice(i, i + BATCH_SIZE);
    console.log(`\n📦 Batch ${Math.floor(i/BATCH_SIZE) + 1}/${Math.ceil(toProcess.length/BATCH_SIZE)} (${batch.length} items)`);
    
    // Process each municipality in batch
    for (const place of batch) {
      console.log(`  🔍 ${place.name}, ${place.state_abbr} (${place.population_2020.toLocaleString()} pop)`);
      
      const orgSearchTerm = formatOrgSearchTerm(place);
      console.log(`     Searching: "${orgSearchTerm}"`);
      
      const contacts = await searchApolloContacts(orgSearchTerm, place.state_abbr);
      
      if (contacts.length > 0) {
        console.log(`     ✅ Found ${contacts.length} contact(s)`);

        place.contacts = contacts;
        place.contact_name = contacts[0].name;
        place.contact_title = contacts[0].title;
        place.contact_email = contacts[0].email;
        place.contact_phone = contacts[0].phone;
        place.contact_source = 'apollo';
        place.contact_enriched_at = new Date().toISOString();

        progress.totalEnriched += contacts.length;
      } else {
        console.log(`     ❌ No contacts found`);
        place.contacts = [];
        place.contact_enriched_at = new Date().toISOString();
        place.contact_source = 'apollo_attempted';
      }
      
      // Mark as processed
      progress.processed.push(place.geoid);
      progress.totalFound = progress.processed.length;
      
      // Save progress after each item
      saveProgress(progress);
    }
    
    // Delay between batches (except after last batch)
    if (i + BATCH_SIZE < toProcess.length) {
      console.log(`\n⏳ Waiting ${DELAY_MS/1000}s before next batch...`);
      await new Promise(resolve => setTimeout(resolve, DELAY_MS));
    }
  }
  
  // Write updated data back to file
  console.log('\n💾 Saving updated data...');
  data.places = places; // Update with modified places
  fs.writeFileSync(DATA_FILE, JSON.stringify(data, null, 2));
  
  console.log('\n✅ Enrichment complete!');
  console.log(`📊 Summary:`);
  console.log(`   Total target municipalities: ${targetPlaces.length}`);
  console.log(`   Processed: ${progress.totalFound}`);
  console.log(`   Enriched with contacts: ${progress.totalEnriched}`);
  console.log(`   Success rate: ${((progress.totalEnriched / progress.totalFound) * 100).toFixed(1)}%`);
  
  // Update metadata
  if (!data.metadata.enrichment) {
    data.metadata.enrichment = {};
  }
  data.metadata.enrichment.last_run = new Date().toISOString();
  data.metadata.enrichment.target_population_band = '50k+';
  data.metadata.enrichment.target_count = targetPlaces.length;
  data.metadata.enrichment.processed_count = progress.totalFound;
  data.metadata.enrichment.enriched_count = progress.totalEnriched;
  data.metadata.enrichment.source = 'Apollo.io People Search API';
  data.metadata.enrichment.title_filters = TITLE_FILTERS;
  
  fs.writeFileSync(DATA_FILE, JSON.stringify(data, null, 2));
  console.log('\n📝 Metadata updated.');
}

// Run with error handling
main().catch(error => {
  console.error('❌ Fatal error:', error);
  process.exit(1);
});